lib/pbench/agent/tool_meister_client.py

```diff
@@ -146,82 +101,55 @@ def publish(self, group, directory, action, args=None):
+        metadata = dict(group=group, directory=str(directory), args=args)
+        self.logger.debug(
+            f"publish state signal for state {action} with metadata: {metadata}"
+        )
         try:
-            num_present = self.redis_server.publish(
-                self.from_client_channel, json.dumps(msg)
+            resp, msgs = self.sig_pub.publish_signal(
+                event=action, tag="from_pbench_client", metadata=metadata, timeout=100
             )
         except Exception:
-            self.logger.exception("Failed to publish client message")
+            self.logger.exception("Failed to publish client signal")
             return 1
         else:
-            self.logger.debug("published %s", self.from_client_channel)
-            if num_present != 1:
-                self.logger.error(
-                    "Failed to publish to the TDS, encountered %d subscribers"
-                    " on the channel",
-                    num_present,
-                )
+            if resp != 0:
+                self.logger.error(f"Missing or bad response from the TDS, {resp}")
                 ret_val = 1
+                for responder, msg in msgs.items():
+                    if msg != "success":
+                        self.logger.warning(
+                            f"TDS responder {responder} reported: {msg}"
+                        )
             else:
                 ret_val = 0
-
-        # Wait for an operational status message from the Tool Data Sink
-        # reporting the combined response of the Tool Meisters.
-        for data in self.to_client_chan.fetch_json(self.logger):
-            try:
-                kind = data["kind"]
-                action_r = data["action"]
-                status = data["status"]
-            except Exception:
-                self.logger.error("unrecognized status payload, %r", data)
-                ret_val = 1
-            else:
-                if kind != "ds":
-                    self.logger.warning("unrecognized kind in payload, %r", data)
-                    ret_val = 1
-                    continue
-                if action_r != action:
-                    self.logger.warning("unrecognized action in payload, %r", data)
-                    ret_val = 1
-                    continue
-                if status != "success":
-                    self.logger.warning(
-                        "Status message not successful: '%s'", status,
-                    )
-                    ret_val = 1
-                break
         return ret_val
 
+    def publish(self, group, directory, action, args=None):
+        """publish a state signal formed from the group, directory,
+        action, and args arguments.
+
+        Returns 0 on success, 1 on failure; logs are also written for any
+        errors encountered.
+        """
+        if action not in api_tm_allowed_actions:
+            return 1
```

Comment:
We still need a check for at least `tm_allowed_actions`, as we want to make sure the signal the user is trying to publish is in the legal event space (it would be caught by state-signals eventually, but it'd be quicker to just catch it b4 any other work). I can switch to just using `tm_allowed_actions` and allowing terminates to also be published through the standard publish method, though. I will also add logging for when we catch an illegal event.